kubernetes/controllers/control/identity_secret_controller.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	log.Info("reconciling secret...", "secret", secretName)
+
+	for _, identity := range identityList.Items {
+		if identity.Status.NeedRenewal {
+			continue
+		}
+		patch := kclient.MergeFrom(identity.DeepCopy())
+		identity.Status.NeedRenewal = true
+		if err := r.kube.Status().Patch(ctx, &identity, patch); err != nil {
+			return ctrl.Result{}, trace.Wrap(err)
+		}
+		log.Info("identity marked for renewal", "identity", identity.Name, "secret", secretName)
+	}
+
+	return ctrl.Result{RequeueAfter: validityDuration}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r identitySecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	predicates := []predicate.Predicate{predicate.Funcs{
+		UpdateFunc: func(evt event.UpdateEvent) bool {
+			secretNew, ok := evt.ObjectNew.(*corev1.Secret)
+			if !ok {
+				return false
+			}
+			secretOld, ok := evt.ObjectOld.(*corev1.Secret)
+			if !ok {
+				return false
+			}
+			return !bytes.Equal(secretNew.Data[IdentityKey], secretOld.Data[IdentityKey])
+		},
+	}}
+	return trace.Wrap(ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}, builder.WithPredicates(predicates...)).
+		Complete(r),
+	)
+}
+
+func (r identitySecretReconciler) verifyIdentity(ctx context.Context, identityStr string) (bool, time.Duration, error) {
+	log := klog.FromContext(ctx)
+	identityFile, err := identityfile.FromString(identityStr)
+	if err != nil {
+		return false, 0, trace.Wrap(err, "failed to parse the identity contents, maybe the secret data has been changed or deleted")
+	}
+	cas, err := r.signer.GetCAs(ctx)
+	if err != nil {
+		return false, 0, trace.Wrap(err, "failed to load CA set")
+	}
+	caSet, err := parseCAs(ctx, cas)
+	if err != nil {
+		return false, 0, trace.Wrap(err, "failed to parse CA set")
+	}
+
+	// Parse the TLS certificate.
+	pemBlock, _ := pem.Decode(identityFile.Certs.TLS)
+	if pemBlock.Type != "CERTIFICATE" {
+		return false, 0, trace.Errorf("failed to parse tls certificate: bad block type %s", pemBlock.Type)
+	}
+	tlsCert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return false, 0, trace.Wrap(err, "failed to parse tls certificate")
+	}
+
+	// Parse the SSH certificate.
+	sshPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(identityFile.Certs.SSH)
+	if err != nil {
+		return false, 0, trace.Wrap(err, "failed to parse ssh certificate")
+	}
+	sshCert, ok := sshPublicKey.(*ssh.Certificate)
+	if !ok {
+		return false, 0, trace.BadParameter("failed to parse ssh certificate: not a certificate")
+	}
+
+	if err := verifyIdentityWithCA(tlsCert, sshCert, caSet.activeTLS, caSet.activeSSH); err != nil {
+		if err := verifyIdentityWithCA(tlsCert, sshCert, caSet.additionalTLS, caSet.additionalSSH); err == nil {
+			log.Info("the identity has been signed with a rotated CA key set, need to refresh")
+			return false, 0, nil
+		}
+		return false, 0, trace.Wrap(err)
+	}
+
+	return true, time.Until(time.Unix(int64(sshCert.ValidBefore), 0)), nil
```

Comment:
Is an identify considered verified outside the certificate's validity period, or has `verifyIdentityWithCA()` already checked that?

Reply:
How is this being tested, by the way?

Reply:
My bad, in current implementation it's impossible to test because `controller-runtime` package doesn't support clock mocking.

I'd rather extract that piece of code and write isolated tests.